File: Array/119. Pascal's Triangle II.py
'''
Method 1:
'''
class Solution(object):
    def getRow(self, rowIndex):
        """
        :type rowIndex: int
        :rtype: List[int]
        """
        ans = []
        ans.append(1)
        for i in range(1, rowIndex+1):
            ans.append(1)
            for j in range(i-1, 0, -1):
                ans[j] = ans[j] + ans[j-1]
        
        return ans
        
'''
Method 2: Recursive approach
'''
Pascal = defaultdict(list)
class Solution(object):
    def getRow(self, rowIndex):
        """
        :type rowIndex: int
        :rtype: List[int]
        """
        if rowIndex == 0:
            Pascal[rowIndex] = [1]
            return [1]
        elif rowIndex == 1:
            Pascal[rowIndex] = [1, 1]
            return [1, 1]
        row = [0]*(rowIndex+1)
        row[0], row[-1] = 1, 1
        for i in range(1,len(row)-1,1):
            if rowIndex-1 not in Pascal:
                self.getRow(rowIndex-1)
            row[i] = Pascal[rowIndex-1][i-1] + Pascal[rowIndex-1][i]
        Pascal[rowIndex] = row
        return row
        

File: Array/119. Pascal's Triangle II.cpp
/*
119. Pascal's Triangle II
Array

Description:
Given an integer rowIndex, return the rowIndexth (0-indexed) row of the Pascal's triangle.
In Pascal's triangle, each number is the sum of the two numbers directly above it as shown:
1
11
121
1331
14641

Example 1:
Input: rowIndex = 3
Output: [1,3,3,1]

Example 2:
Input: rowIndex = 0
Output: [1]

Example 3:
Input: rowIndex = 1
Output: [1,1]

Similar Question:
Pascal's Triangle - Easy
*/


class Solution {
public:
    vector<int> getRow(int rowIndex) {
        vector<int> result;
        result.resize(rowIndex + 1);
        result[0] = 1;
        result[rowIndex] = 1;
        for (int i = 1; i <= rowIndex; ++i) {
            result[i] = 1;
            for (int j = i - 1; j > 0; --j) {
                result[j] = result[j] + result[j - 1];
            }
        }
        return result;
      }
};
